scipy/_lib/_array_api_override.py

```diff
@@ -99,24 +97,66 @@ def array_namespace(*arrays: Array) -> ModuleType:
 
     Notes
     -----
-    Thin wrapper around `array_api_compat.array_namespace`.
+    Wrapper around `array_api_compat.array_namespace`.
+
+    1. Check for the global switch `SCIPY_ARRAY_API`. If disabled, just
+       return array_api_compat.numpy namespace and skip all compliance checks.
+
+    2. Check for known-bad array classes.
+       The following subclasses are not supported and raise and error:
 
-    1. Check for the global switch: SCIPY_ARRAY_API.
-    2. `_compliance_scipy` raise exceptions on known-bad subclasses. See
-       its definition for more details.
+       - `numpy.ma.MaskedArray`
+       - `numpy.matrix`
+       - NumPy arrays which do not have a boolean or numerical dtype
+       - `scipy.sparse` arrays
 
-    When the global switch is False, it defaults to the `numpy` namespace.
-    In that case, there is no compliance check. This is a convenience to
-    ease the adoption. Otherwise, arrays must comply with the new rules.
+    3. Coerce array-likes to NumPy arrays and check their dtype.
+       Note that non-scalar array-likes can't be mixed with non-NumPy Array
+       API objects; e.g.
+
+       - `array_namespace([1, 2])` returns NumPy namespace;
+       - `array_namespace(np.asarray([1, 2], [3, 4])` returns NumPy namespace;
+       - `array_namespace(cp.asarray([1, 2], [3, 4])` raises an error.
     """
     if not SCIPY_ARRAY_API:
         # here we could wrap the namespace if needed
         return np_compat
 
-    api_arrays = list(_compliance_scipy(arrays))
-    # In case of a mix of array API compliant arrays and scalars, return
-    # the array API namespace. If there are only ArrayLikes (e.g. lists),
-    # return NumPy (wrapped by array-api-compat).
-    if api_arrays:
-        return array_api_compat.array_namespace(*api_arrays)
-    return np_compat
+    numpy_arrays = []
+    api_arrays = []
+
+    for array in arrays:
+        arr_info = _validate_array_cls(type(array))
+        if arr_info is _ArrayClsInfo.skip:
+            pass
+        elif arr_info is _ArrayClsInfo.is_numpy:
+            dtype_info = _validate_numpy_dtype(array.dtype)
+            if dtype_info is _DTypeInfo.maybe_jax:
+                if is_jax_array(array):
+                    api_arrays.append(array)  # JAX zero gradient array
+                else:
+                    # void dtype, but not a JAX zero gradient array
+                    raise TypeError(f"An argument has dtype `{array.dtype!r}`; "
+                        f"only boolean and numerical dtypes are supported.")
+            else:
+                numpy_arrays.append(array)
+        elif arr_info is _ArrayClsInfo.no_info and is_array_api_obj(array):
+            api_arrays.append(array)
+        else:
+            # list, tuple, or arbitrary object
```

Comment:
we could drastically speed up the use case of ArrayLikes by pushing this validation to `_asarray`. However, that would require sanitizing all calls to `xp.asarray` throughout the codebase, so I'm leaving it to a follow-up.